### DISCUSSION TITLE
Password recovery problem: getting INITIAL_SESSION instead of SIGNED_IN / PASSWORD_RECOVERY events

### DISCUSSION THREAD

**Discussion by alexbriannaughton** [supabase/supabase#18059] [Questions]
Hello!

I'm trying to add password recovery to an Expo Router app using Supabase's built in flow. However, whenever I click the link that's sent to my email, the link simply takes me to my redirect link without the correctly emitted events. When I log the event within my onAuthStateChange listener, I'm only getting an INITIAL_SESSION event instead of the SIGNED_IN and PASSWORD_RECOVERY events described in the docs.

Below is a link to a repo of a minimal Expo Router Project with Supabase Auth that reproduces the issue on my end. Any thoughts on what I'm doing wrong here?

https://github.com/alexbriannaughton/expo-router-with-supabase-auth

**Comment by s0h311** [comment]
Hi, @alexbriannaughton, this is how I do it:

```JavaScript
await supabase.auth.resetPasswordForEmail(email.value, {
  redirectTo: 'https://example.com/setNewPassword',
})
```

And before the `/setNewPassword` page mounts
```JavaScript
supabase.auth.onAuthStateChange((event, _session) => {
  if (event == 'PASSWORD_RECOVERY') {
    // do something
  }  
})
```

**Reply by s0h311** [reply]
@alexbriannaughton I'm not familiar with expo but maybe this is the reason:

> Currently, onAuthStateChange() does not work across tabs. For instance, in the case of a password reset flow, the original tab which requested for the password reset link will not receive the SIGNED_IN and PASSWORD_RECOVERY event when the user clicks on the link.

See https://supabase.com/docs/reference/javascript/auth-onauthstatechange

**Reply by alexbriannaughton** [reply]
@s0h311 Thanks for reaching out. This is actually pretty well the same code that I already have. Good idea with the new tab, but I seem to get the wrong event in a new tab as well.

**Reply by s0h311** [reply]
@alexbriannaughton If I'm remembering it correctly, the reset link also signs in the user, so the `INITIAL_SESSION` kind of makes sense becuase in your `useEffect` you are redirecting the user: 

```JS
supabase.auth.getSession().then(({ data: { session } }) => {
  if (session) router.replace("/(tabs)/home");
})
```

My guess is without redirection the `PASSWORD_RECOVERY` event fires up.

**Reply by alexbriannaughton** [reply]
Unfortunately, the behavior is the same when I comment out that line of code!

**Comment by bhimeshchauhan** [comment]
@alexbriannaughton were you able to figure this one out?

**Reply by alexbriannaughton** [reply]
i wasnt unfortunately 😢
